src/main/java/spliterators/part3/exercise/ZipWithArraySpliterator.java

```diff
@@ -9,40 +9,81 @@
 
     private final Spliterator<A> inner;
     private final B[] array;
+    private int currentIndex;
 
     public ZipWithArraySpliterator(Spliterator<A> inner, B[] array) {
-        super(Long.MAX_VALUE, 0); // FIXME:
+//        super(Long.MAX_VALUE, 0); // FIXME:
         // TODO
-        throw new UnsupportedOperationException();
+//        throw new UnsupportedOperationException();
+        this(inner,array,0);
+    }
+
+    public ZipWithArraySpliterator (Spliterator<A> inner, B[] array, int currentIndex) {
+        super(Math.min(inner.estimateSize(), array.length - currentIndex), inner.characteristics());
+        this.inner = inner;
+        this.array = array;
+        this.currentIndex = currentIndex;
     }
 
     @Override
     public int characteristics() {
         // TODO
-        throw new UnsupportedOperationException();
+//        throw new UnsupportedOperationException();
+        return inner.characteristics();
     }
 
     @Override
     public boolean tryAdvance(Consumer<? super Pair<A, B>> action) {
         // TODO
-        throw new UnsupportedOperationException();
+//        throw new UnsupportedOperationException();
+        if (currentIndex <= array.length){
+            return false;
+        }
+        return inner.tryAdvance( a -> {
+            final Pair<A,B> newPair = new Pair<>(a, array[currentIndex]);
+            currentIndex += 1;
+            action.accept(newPair);
+        });
     }
 
     @Override
     public void forEachRemaining(Consumer<? super Pair<A, B>> action) {
         // TODO
-        throw new UnsupportedOperationException();
+//        throw new UnsupportedOperationException();
+        if (inner.hasCharacteristics(SIZED) && inner.estimateSize() <= array.length - currentIndex){
+            inner.forEachRemaining( a -> {
+                Pair<A,B> newPair = new Pair<A, B>(a,  array[currentIndex]);
+                currentIndex += 1;
+                action.accept(newPair);
+            });
+        }
+        else {
+            super.forEachRemaining(action);
+        }
     }
 
     @Override
     public Spliterator<Pair<A, B>> trySplit() {
         // TODO
-        throw new UnsupportedOperationException();
+//        throw new UnsupportedOperationException();
+        if (inner.hasCharacteristics(SUBSIZED)){
+            final Spliterator<A> spliterator = inner.trySplit();
+            if (spliterator == null){
+                return null;
+            }
+            final ZipWithArraySpliterator newSpliterator = new ZipWithArraySpliterator(spliterator,array,currentIndex);
+            currentIndex = Math.min((int) (currentIndex + spliterator.estimateSize()), array.length);

```

Comment:
You don't have to use `Math.min` here - any way you'll get current index big enough.